pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+import sys
+from typing import (
+    Any,
+    Generic,
+    Protocol,
+    TypeVar,
+    Union,
+    overload,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    Timedelta,
+    Timestamp,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+VALID_CLOSED: frozenset
+
+OrderableScalarT = TypeVar("OrderableScalarT", int, float)
+OrderableTimesT = TypeVar("OrderableTimesT", Timestamp, Timedelta)
+OrderableT = TypeVar("OrderableT", int, float, Timestamp, Timedelta)
+
+class IntervalMixinProtocol(Protocol): ...
+
+class _LengthDescriptor:
+    @overload
+    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
+    @overload
+    def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> Timedelta: ...
+
+class IntervalMixin:
+    @property
+    def closed_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def closed_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def mid(self: IntervalMixinProtocol) -> float: ...
+    @property
+    def is_empty(self: IntervalMixinProtocol) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...): ...
+
+class Interval(IntervalMixin, Generic[OrderableT]):
+    @property
+    def left(self: Interval[OrderableT]) -> OrderableT: ...
+    @property
+    def right(self: Interval[OrderableT]) -> OrderableT: ...
+    @property
+    def closed(self) -> str: ...
+    def __init__(
+        self,
+        left: OrderableT,
+        right: OrderableT,
+        closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
```

Comment:
remove `str`, wouldn't need the literals otherwise.

Reply:
OK